Fix relocation to avoid overwriting merged constant strings

trws opened the pull request:
This is the fix for issues observed in #31739, and causing failures to build the mesa package possibly since last September.  The observed symptom of the problem was that various floating point constants in `src/util/format_srgb.c` in mesa were producing syntax errors in exactly one configuration, of two, on ARM graviton nodes on AWS and absolutely nowhere else.  Specifically numbers that should have been `7.6052450e-01` were `7.6052450/-01` instead.  Thanks to some poking from @tgamblin, we determined that source file was generated by python, and tracked down the issue to python's `%` formatting:

```python-console
>>> i = 7.6052450e-01
>>> i
0.7605245
>>> "%.7Ef" % i
'7.6052450E-01f'
>>> "%.7ef" % i
'7.6052450/-01f'
```

Note that `E` formatting works correctly, but `e` does not.  The formatting logic comes from `pystrtod.c` in cpython, and uses [a pair of global string arrays](https://github.com/python/cpython/blob/main/Python/pystrtod.c#L966-L975) to differentiate between upper- and lower-case strings for floating point formatting.  The `lc_float_strings` table in `libpython.so` was corrupted:

```
(gdb) p lc_float_strings
$1 = {0x2e1528 "inf", 0x2f5920 "nan", 0x2f6f88 "/"}
```

More interesting was how, the original version that had it correct, when searching back from the start address of the "e" string showed that the linker re-used the "e" *at the end of the spack package prefix* as the string constant for e:

```
(gdb) p lc_float_strings
$1 = {0x2e1528 "inf", 0x2f5920 "nan", 0x2f6f88 "e"}
$6 = 0x2f6f50 "/gcc-7.3.1/python-3.9.13-g3acg4n4bzr2o5wuukebasdfkjym2x5e"
```

This happens whenever `-fmerge-constants` is active in gcc, clang etc. which it is for `-O1` and higher, so since approximately September of 2021 we have likely shipped uniquely broken binaries of python whenever the hash happened to end in "e".  The reason is that our relocation logic, originally designed to pad the new prefix with `os.sep` characters, was accidentally refactored to *suffix* with them instead.

After much discussion, we determined that ensuring splicing preserved a suffix, shortening when possible, and prefix padding with separators when not would be the best course of action.  This PR implements the hash suffix preservation for splicing with a length of seven, shortens the prefix and terminates with a null without overwriting the rest when the prefix is immediately followed by a null byte, and after that prefix pads whatever matches remain.

If we find that the prefix padding is a usability problem, we can shorten cases where the null terminator is further away as long as the intervening space doesn't contain another copy of the prefix, but there are risks, so we're starting with this.

trws committed:
fix relocation order to preserve the suffix

trws committed:
update relocation tests

trws committed:
relocation seems to be working, rewiring is completely broken

trws committed:
implement 3-piece relocation

trws committed:
style fixes

trws committed:
Merge remote-tracking branch 'upstream/develop' into fix-relocate

haampie commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -1588,7 +1588,7 @@ def relocate_package(spec, allow_root):
     prefix_to_prefix_bin[old_prefix] = new_prefix
     prefix_to_prefix_text[old_layout_root] = new_layout_root
-    prefix_to_prefix_bin[old_layout_root] = new_layout_root
+    # prefix_to_prefix_bin[old_layout_root] = new_layout_root
```
I found this one odd too. Was it intended as a generic fallback? In that case it belongs below the loop over spec prefixes

trws replied:
I'm honestly not sure, all the tests pass without it (though I'm still fixing up the rework for a single regex version).

haampie replied:
I'll put this back in place, but after the prefix loop

trws committed:
finish fixing relocation for single regex style and reinstate suffix checks

trws committed:
use sh rather than bash for generated scripts in conftest

trws committed:
Merge remote-tracking branch 'upstream/develop' into fix-relocate

large rework to tolerate moving hello_world into binary_with_rpaths

trws committed:
fix style

trws commented:
Ok, this should be good to go pending comments.  I put the check for non-matching suffixes back in at a length of 7. This should not happen, but if someone puts something other than hash last in their projection and tries to relocate, they'll get an explanation rather than crazy failures.

trws commented:
This however is not great news, apparently bootstrap in CI relocates to *longer paths*? https://github.com/spack/spack/actions/runs/3374819566/jobs/5600910914#step:4:433

haampie committed:
Consolidate DAG traversal in traverse.py, support DFS/BFS (#33406)

This PR introduces breadth-first traversal, and moves depth-first traversal
logic out of Spec's member functions, into `traverse.py`.

It introduces a high-level API with three main methods:

```python
spack.traverse.traverse_edges(specs, kwargs...)
spack.traverse.traverse_nodes(specs, kwags...)
spack.traverse.traverse_tree(specs, kwargs...)
```

with the usual `root`, `order`, `cover`, `direction`, `deptype`, `depth`, `key`,
`visited` kwargs for the first two.

What's new is that `order="breadth"` is added for breadth-first traversal.

The lower level API is not exported, but is certainly useful for advanced use
cases. The lower level API includes visitor classes for direction reversal and
edge pruning, which can be used to create more advanced traversal methods,
especially useful when the `deptype` is not constant but depends on the node
or depth. 

---

There's a couple nice use-cases for breadth-first traversal:

- Sometimes roots have to be handled differently (e.g. follow build edges of
  roots but not of deps). BFS ensures that root nodes are always discovered at
  depth 0, instead of at any depth > 1 as a dep of another root.
- When printing a tree, it would be nice to reduce indent levels so it fits in the 
  terminal, and ensure that e.g. `zlib` is not printed at indent level 10 as a 
  dependency of a build dep of a build dep -- rather if it's a direct dep of my
  package, I wanna see it at depth 1. This basically requires one breadth-first
  traversal to construct a tree, which can then be printed with depth-first traversal.
- In environments in general, it's sometimes inconvenient to have a double
  loop: first over the roots then over each root's deps, and maintain your own
  `visited` set outside. With BFS, you can simply init the queue with the
  environment root specs and it Just Works. [Example here](https://github.com/spack/spack/blob/3ec73046995d9504d6e135f564f1370cfe31ba34/lib/spack/spack/environment/environment.py#L1815-L1816)

tgamblin commented:
@trws can you rebase this on `develop` and update the PR message with what we talked about in the meeting Tuesday (basically that you're not handling shortening when there are multiple instances of the prefix/hash)?

tgamblin committed:
tests: fix group membership check in sbang tests. (#33658)

Fixes an issue on the RHEL8 UBI container where this test would fail because `gr_mem`
was empty for every entry in the `grp` DB.

You have to check *both* the `pwd` database (which has primary groups) and `grp` (which
has other gorups) to do this correctly.

- [x] update `llnl.util.filesystem.group_ids()` to do this
- [x] use it in the `sbang` test

fspiga committed:
armpl-gcc: Pull RHEL8 package when OS is Rocky8 (#33641)

trws committed:
only raise if the path would actually be replaced

trws commented:
I'm merging rather than rebasing at the moment, can rebase to clean this up in a bit once we work out the  bootstrap situation.  Updating the description momentarily.

trws committed:
allow g++ to be somewhere other than /usr/bin

trws committed:
allow different suffix as long as the prefix is not followed by null

trws commented:
This should work with rewiring again now.  There's a case that we neglected to consider in rewiring, where the package being rewired needs to replace the paths *of other packages* that are not having their hashes replaced and thus we can't frankenhash.  For right now, I have the check failing these only if the prefix is followed immediately by a null byte, because that's the only case when that suffix could have been reused.  For all the cases we test in the repository, that works out fine, but I have no way to predict how often that will be a problem for rewiring in the wild.  We should probably consider solutions to that for the future.

alalazo commented:
@spackbot rebuild everything

haampie commented on lib/spack/spack/relocate.py:
```diff
@@ -467,39 +480,66 @@ def _replace_prefix_bin(filename, byte_prefixes):
-        for replacement, offset in replacements_at_offset:
-            f.seek(offset)
-            f.write(replacement)
+            if len(new) > len(old) or (terminated and new[-7:] != old[-7:]):
```
I'm confused, why require a common suffix? You get `len(old) - len(new) - 1` unused bytes after replacement + new null byte, so a new suffix is possible when that's >= 7. And if you require a common suffix, you can just as well pad with leading `/`, like in the non-null terminated case.

Right now if the the new length is 1 byte shorter and the suffix has 7 bytes in common, it's still broken. E.g. `/abc/1234567` -> `/ab/1234567` would turn `/abc/1234567\0` into `/ab/1234567\0\0`, which is incorrect.

trws replied:
Darn it, you're right, the check isn't quite right anymore.

The suffix case is mainly to deal with same-length or shorter than safe width difference prefixes when terminated.  We can't do anything if the suffixes aren't the same.  It also impacts padding in similar cases. Probably it should be:
* if len(new) - len(old)  > 7 and terminated: ok, shorten
* if len(new) - len(old)  > 7 and not terminated: ok, pad
* if len(new) - len(old)  < 7 and terminated and matching: ok, pad
* if len(new) - len(old)  < 7 and terminated and not matching: 💥 
* if len(new) - len(old)  < 7 and not terminated and matching: ok, pad
* if len(new) - len(old)  < 7 and not terminated and not matching: may be ok to pad, depends on length of suffix

Need to add tests so we cover all of these.  Honestly kinda want to fuzz the 💩 out of this too.

haampie replied:
I'm happy to push a few commits when you're on holidays ;p

haampie replied:
How about

```python
if len(new) > len(old):
  raise NoInplaceUpdatePossible()
elif not null_terminated or old[-N:] == new[-N:]:
  replacement = b"/" * (len(old) - len(new)) + new
elif len(new) + N < old:  # needs < cause of \0 after new
  replacement = new + b"\0"
else:
  raise AliasingRiskIssues()
```

haampie replied:
I added tons of tests now, and also introduced a lookahead of 7 bytes for a terminating null instead of direct nulls. That covers the case of replacing e.g. `/pkg-<hash>` in `/pkg-<hash>/\0` when the replacement has a different suffix (for some reason).

trws replied:
Awesome! I'm floored @haampie, this looks fabulous.  I threw in one suggestion to fix an indentation bug but pending a defensive recheck on gitlab I think this is GTG.

haampie committed:
Merge branch 'develop' into fix-relocate

haampie committed:
Deal with all cases [null terminated in lookahead window, not null terminated] x [common suffix, not common suffix] x [sufficiently short, unsufficiently short]

haampie committed:
restore prefix->prefix of old_layout_root

haampie committed:
docs

haampie committed:
typo

haampie committed:
typo

haampie committed:
typo

haampie committed:
remove redundant tmpdir

haampie committed:
make flake8 happy

trws commented on lib/spack/spack/relocate.py:
```diff
@@ -463,43 +470,116 @@ def _replace_prefix_text(filename, compiled_prefixes):
-        byte_prefixes (OrderedDict): OrderedDictionary where the keys are
-        binary strings of the old prefixes and the values are the new
-        binary prefixes
+        byte_prefixes (OrderedDict): ordered dictionary where the keys are
+            bytes representing the old prefixes and the values are the new
+        prefixes (all bytes utf-8 encoded)
```
```suggestion
        byte_prefixes (OrderedDict): ordered dictionary where the keys are
            bytes representing the old prefixes and the values are the new
            prefixes (all bytes utf-8 encoded)
```

trws committed:
Update lib/spack/spack/relocate.py

trws committed:
Merge branch 'develop' into fix-relocate

trws commented:
@spackbot rebuild everything

haampie commented:
Rebuild everything worked except for the tau shasum which was fixed on develop 34724cae87623a36de0c34e6e6bba55e969c4a96
